tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,693 @@
+        with ops.control_dependencies([v_op]):
+          for global_var, v in zip(global_vars, v_vars):
+            model_reassign_ops.append(state_ops.assign_sub(global_var, v))
+          g_update_op = control_flow_ops.group(*(model_reassign_ops))
+        with ops.control_dependencies([g_update_op]):
+          for n_var, global_var, v in zip(nesterov_vars, global_vars, v_vars):
+            momentum = math_ops.multiply(self._block_momentum_rate, v)
+            nesterov = math_ops.subtract(global_var, momentum)
+            nesterov_reassign_ops.append(state_ops.assign(n_var, nesterov))
+          nesterov_reassign_ops.append(state_ops.assign_add(global_step, 1))
+          update_op = control_flow_ops.group(*(nesterov_reassign_ops))
+
+
+      self._local_step = self._local_step
+      self._train_ops = train_ops
+      # Create token queue.
+      with ops.device(global_step.device), ops.name_scope(""):
+        sync_token_queue = (
+            data_flow_ops.FIFOQueue(-1,
+                                    global_step.dtype.base_dtype,
+                                    shapes=(),
+                                    name="sync_token_q",
+                                    shared_name="sync_token_q"))
+        self._sync_token_queue = sync_token_queue
+
+        # dummy_queue is passed to the queue runner. Don't use the real queues
+        # because the queue runner doesn't automatically reopen it once it
+        # closed queues in PS devices.
+        dummy_queue = (
+            data_flow_ops.FIFOQueue(1,
+                                    types_pb2.DT_INT32,
+                                    shapes=(),
+                                    name="dummy_queue",
+                                    shared_name="dummy_queue"))
+
+      with ops.device(global_step.device), ops.name_scope(""):
+        # Replicas have to wait until they can get a token from the token queue.
+        with ops.control_dependencies(train_ops):
+          token = sync_token_queue.dequeue()
+        train_op = state_ops.assign(self._local_step, token)
+
+        with ops.control_dependencies([update_op]):
+          # Sync_op needs to insert tokens to the token queue at the end of the
+          # step so the replicas can fetch them to start the next step.
+          tokens = array_ops.fill([self._tokens_per_step], global_step)
+          sync_op = sync_token_queue.enqueue_many((tokens,))
+
+        self._chief_queue_runner = queue_runner.QueueRunner(dummy_queue,
+                                                            [sync_op])
+      for accum, dev in self._accumulator_list:
+        with ops.device(dev):
+          chief_init_ops.append(
+              accum.set_global_step(
+                  global_step, name="SetGlobalStep"))
+      self.chief_init_op = control_flow_ops.group(*(chief_init_ops))
+      self._gradients_applied = True
+      return train_op
+
+  def apply_refresh_local_vars(self, local_vars, global_vars):
```

Comment:
Maybe this should be "assign" instead of "append"?  (And private methods should have a leading underscore.)